### PR TITLE
Cell statusbar API

### LINKED ISSUE #105809: Notebook cell status bar API
We have a proposal for cell status bar items but we should revisit its design. The current proposal is modelled closely after the existing global status bar API. However the global status bar and cell status bars have a very different lifecycle: the global status bar is always there and shares the lifecycle of the editor. Cells on the other hand are usually short-lived, often not visible (not scrolled in view, hidden editor etc pp). 

This is what's needed to manage the status bar for cells: https://github.com/microsoft/vscode-github-issue-notebooks/blob/edf85a2e9db0d28166526c844e479724434209b6/src/extension/notebookProvider.ts#L156-L182 and it shows a couple of weaknesses of the design

* it has much semantic overlap with commands/menus
* it listens to output changes of any notebook (no matter if visible or not), e.g it actively needs to manage status bar item which is often error-prone
* it doesn't know anything about the lifecycle because the workbench will dispose the object the extension has created. That's generally fishy because only object-creator should call dispose.

So, I think we should first justify why context keys and commands aren't sufficient and then we should turn this into a provider model. E.g have a `NotebookCellStatusBarItemProvider` which we ask for cell's items, e.g

```ts
class NotebookCellStatusBarItem {
  command: Command;
  text: string;
}

interface NotebookCellStatusBarItemProvider {
  // provider has changed, 
  onDidChange?: Event<void>;

  provideCellStatusBarItem(cell: NotebookCell): ProviderResult<NotebookCellStatusBarItem>;
}
```

We decide when to call `provideCellStatusBarItem`, like on output change etc, and it will return short-lived data-only objects that we manage.

### PR DESCRIPTION
**Pull request by roblourens**
Fix #105809

Much better implementation of the cell statusbar API.

Open questions/todos

- Do extensions need to show items that are only visible when the cell is active? Should I just call the providers again when the cell becomes active?
- Need to debounce statusbar updates when scrolling? 
- Any other cases when providers should be called? The only case when the provider is canceled is when the item scrolls out of view during an update, is that right?
- The service is kind of pointless here, would it be better to just expose a list of providers, basically a provider registry?
- Get other internal things onto this API, and remove `statusMessage` metadata
- Fix bugs like https://github.com/microsoft/vscode/issues/115099 and https://github.com/microsoft/vscode/issues/112772

### CODE REVIEW

File: src/vs/workbench/contrib/notebook/browser/contrib/statusBar/statusBarProviders.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import { Registry } from 'vs/platform/registry/common/platform';
+import { Extensions as WorkbenchExtensions, IWorkbenchContributionsRegistry } from 'vs/workbench/common/contributions';
+import { INotebookCellStatusBarService } from 'vs/workbench/contrib/notebook/common/notebookCellStatusBarService';
+import { CellStatusbarAlignment, INotebookCellStatusBarItem, INotebookCellStatusBarItemList, INotebookCellStatusBarItemProvider, INotebookDocumentFilter } from 'vs/workbench/contrib/notebook/common/notebookCommon';
+import { INotebookService } from 'vs/workbench/contrib/notebook/common/notebookService';
+import { LifecyclePhase } from 'vs/workbench/services/lifecycle/common/lifecycle';
+
+class CellStatusBarPlaceholderProvider implements INotebookCellStatusBarItemProvider {
+	readonly selector: INotebookDocumentFilter = {
+		filenamePattern: '**/*'
```
Comment (roblourens) [inline src/vs/workbench/contrib/notebook/browser/contrib/statusBar/statusBarProviders.ts:21]: Is there a better way to do this?

File: src/vs/workbench/contrib/notebook/browser/contrib/statusBar/cellStatusBar.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+	private _updateVisibleCells(): void {
+		const vm = this._notebookEditor.viewModel;
+		if (!vm) {
+			return;
+		}
+
+		const newVisibleCells = new Set<number>();
+		this._notebookEditor.visibleRanges.forEach(range => {
+			const cells = this._notebookEditor.getCells({ start: range.start, end: range.end + 1 });
```
Comment (roblourens) [inline src/vs/workbench/contrib/notebook/browser/contrib/statusBar/cellStatusBar.ts:47]: Is there an existing way to do this for a model range? Added `getCells`

**Review by rebornix — COMMENTED** [review]
I added a simple notebook cell status bar items provider like

```ts
	vscode.notebook.registerNotebookCellStatusBarItemProvider({
		viewType: 'mimetype-test'
	}, {
		provideCellStatusBarItems: (cell, token) => {
			return [{
				alignment: vscode.NotebookCellStatusBarAlignment.Left,
				command: undefined,
				text: `${cell.outputs.length} outputs`,
				tooltip: `${cell.outputs.length} outputs`
			}];
		}
	})
```

and after scrolling up and down a couple of times. I'm seeing duplicated status bar items.

image

image

**Comment by rebornix** [comment]
> Need to debounce statusbar updates when scrolling?

It would be great if we debounce it a bit as while scrolling / cell resizing (as they have dynamic height), a cell can be created, disposed and then re-created within a short period.

File: src/vs/workbench/contrib/notebook/browser/viewModel/baseCellViewModel.ts [hunk 3]
```diff
@@ -99,7 +99,11 @@ export abstract class BaseCellViewModel extends Disposable {
 		id?: string;
 		options: model.IModelDeltaDecoration;
 	}>();
-	private _lastDecorationId: number = 0;
+	private _lastItemId: number = 0;
```
[the remark this answers is not in the document]
Comment (roblourens) [reply src/vs/workbench/contrib/notebook/browser/viewModel/baseCellViewModel.ts:102]: I figured I'd reuse it, but it doesn't matter, I can keep it separate

**Comment by roblourens** [comment]
> a cell can be created, disposed and then re-created within a short period.

What do you mean?

The downside of delaying it is that you will see cells without status bar items briefly, then the status bar items will blink into existence, and I found it annoying when testing.

**Comment by roblourens** [comment]
Fixed the issue with dupe items showing up, I thought a cancel token would be canceled on dispose by default.

File: src/vs/vscode.proposed.d.ts [hunk 4]
```diff
@@ -1769,32 +1769,24 @@ declare module 'vscode' {
 		Right = 2
 	}
 
-	export interface NotebookCellStatusBarItem {
-		readonly cell: NotebookCell;
+	export class NotebookCellStatusBarItem {
+		readonly text: string;
 		readonly alignment: NotebookCellStatusBarAlignment;
+		readonly command: string | Command | undefined;
```
Comment (jrieken) [inline src/vs/vscode.proposed.d.ts:1775]: mix of `?` and ` | undefined`

File: src/vs/vscode.proposed.d.ts [hunk 5]
```diff
@@ -1769,32 +1769,24 @@ declare module 'vscode' {
...
-		show(): void;
-		hide(): void;
-		dispose(): void;
+		readonly accessibilityInformation?: AccessibilityInformation;
+
+		constructor(text: string, alignment: NotebookCellStatusBarAlignment, command?: string | Command, tooltip?: string, priority?: number, accessibilityInformation?: AccessibilityInformation);
+	}
+
+	interface NotebookCellStatusBarItemProvider {
+		onDidChangeCellStatusBarItems?: Event<void>;
```
Comment (jrieken) [inline src/vs/vscode.proposed.d.ts:1784]: May also allow to send the event per cell? Like `Event<void | NotebookCell>`

File: src/vs/workbench/contrib/notebook/browser/notebookCellStatusBarServiceImpl.ts [hunk 6]
```diff
@@ -3,51 +3,46 @@
...
+		this._onDidChangeProviders.fire();
 
-		return {
-			dispose: () => {
-				this.removeEntry(entry);
-			}
-		};
+		return toDisposable(() => {
+			changeListener?.dispose();
+			const idx = this._providers.findIndex(p => p === provider);
```
Comment (jrieken) [inline src/vs/workbench/contrib/notebook/browser/notebookCellStatusBarServiceImpl.ts:38]: fyi - we have `LinkedList<T>` which makes removal simpler

File: src/vs/workbench/contrib/notebook/browser/notebookCellStatusBarServiceImpl.ts [hunk 7]
```diff
@@ -3,51 +3,46 @@
...
 	}
 
-	getEntries(cell: URI): INotebookCellStatusBarEntry[] {
-		const existingEntries = this._entries.get(cell);
-		return existingEntries ?
-			Array.from(existingEntries.values()) :
-			[];
+	async getStatusBarItemsForCell(docUri: URI, cellIndex: number, viewType: string, token: CancellationToken): Promise<INotebookCellStatusBarItemList[]> {
+		const providers = this._providers.filter(p => notebookDocumentFilterMatch(p.selector, viewType, docUri));
+		return await Promise.all(providers.map(p => p.provideCellStatusBarItems(docUri, cellIndex, token)));
```
Comment (jrieken) [inline src/vs/workbench/contrib/notebook/browser/notebookCellStatusBarServiceImpl.ts:45]: I'd be more graceful and have try-catch for `provide`-invocations. With that you prevent total failure in the presence of one misbehaving provider

File: src/vs/workbench/api/browser/mainThreadNotebook.ts [hunk 8]
```diff
@@ -196,6 +195,53 @@ export class MainThreadNotebooks implements MainThreadNotebookShape {
...
+
+		const disposable = this._cellStatusBarService.registerCellStatusBarItemProvider(provider);
+		this._notebookCellStatusBarRegistrations.set(handle, disposable);
+	}
+
+	async $unregisterNotebookCellStatusBarItemProvider(handle: number, eventHandle: number | undefined): Promise<void> {
+		const unregisterThing = (handle: number) => {
+			const entry = this._notebookCellStatusBarRegistrations.get(handle);
+			if (entry) {
+				this._notebookCellStatusBarRegistrations.delete(handle);
```
Comment (jrieken) [inline src/vs/workbench/api/browser/mainThreadNotebook.ts:236]: missing dispose?

File: src/vs/workbench/contrib/notebook/browser/contrib/statusBar/statusBarProviders.ts (hunk 1 above)
Comment (jrieken) [reply src/vs/workbench/contrib/notebook/browser/contrib/statusBar/statusBarProviders.ts:21]: Try `*` as viewType or the whole things, like `readonly selector: INotebookDocumentFilter = '*'`

**Review by jrieken — APPROVED** [review]
Only skimmed over the view-parts but what I have seens looks good. Thanks

**Comment by jrieken** [comment]
> The service is kind of pointless here, would it be better to just expose a list of providers, basically a provider registry?

Services are always better than singletons, tho that's all they are: better singletons

File: src/vs/workbench/contrib/notebook/browser/contrib/statusBar/statusBarProviders.ts (hunk 1 above)
Comment (roblourens) [reply src/vs/workbench/contrib/notebook/browser/contrib/statusBar/statusBarProviders.ts:21]: viewType doesn't actually work that way unfortunately, dunno if it should or not

**Pull request merged**